[Actors] Reentrancy discussion and proposal (#38)
DougGregor/swift-evolution: proposals/nnnn-actors.md

Currently, this proposal takes the aggressive approach of assuming _all_ actors are reentrant and not providing ways to opt out of this behavior. This section aims to highlight issues, benefits and tradeoffs with this approach as well as non-reentrancy in general. The goal of discussing these reentrancy issues, is to arrive at a design that is developer friendly, non-surprising, and delivers more completely on Swift's promise of _safe_ and _sane_ concurrency by default thanks to the use of actors.

#### Reentrant actors: "Interleaving" execution

Reentrancy means that execution of asynchronous actor-isolated functions may "interleave" at suspension points, leading to increased complexity in programming with such actors, as every suspension point must be carefully inspected if the code _after_ it depends on some invariants that could have changed before it suspended.

> Interleaving executions still respect the actor's "single-threaded illusion"–i.e. no two functions will ever execute *concurrently* on any given actor–however they may _interleave_ at suspension points. In broad terms this means that reentrant actors are "thread-safe" but are not "data race safe," meaning it is not possible to observe any

Reviewer: Unfinished thought

Reviewer: Also I'm not sure I like calling this a "data race".

Author: Hm yeah data race sounds indeed a bit more like the low level one... Perhaps "high-level race free" to somehow point at that it's about high level invariants rather low level synchonization issues?

Author: reworded as:

> In broad terms this means that reentrant actors are _thread-safe_ but are automatically protecting from the "high level" kinds of races that may still occur, potentially invalidating invariants upon which an executing asynchronous function may be relying on.

To further clarify the implications of this, let us consider the following actor, which thinks of an idea and then returns it, after telling its friend about it.

```swift
// reentrant actor
actor class Person {
  let friend: Friend
  
  // actor-isolated opinion
  var opinion: Judgement = .noIdea

  func thinkOfGoodIdea() async -> Decision {
    opinion = .goodIdea                    // <1>
    await tell(friend, myOpinion: opinion) // <2>
    return opinion // 🤨                   // <3>
  }

  func thinkOfBadIdea() async -> Decision {
    opinion = .badIdea                     // <4>
    await tell(friend, myOpinion: opinion) // <4>
    return opinion // 🤨                   // <6>
  }
}
```

> Reentrant code is notoriously difficult to program with–one is protected from low level data-races, however the higher level semantic races may still happen. In the examples shown, the states are small enough that they are simple to fit in one's hear so they do not appear as tricky, however in the real world races can be dauntingly hard to debug.

> In the example above the `Person` can think of a good or bad idea, shares that opinion with a friend, and returns that opinion that it stored. Since the actor is reentrant this code is **wrong** and will return an _arbitrary opinion_ if the actor begins to think of a few ideas at the same time.

Reviewer: I mean, this is really weird code.  Why is `opinion` actor state to begin with?

Author: The example is silly / oversimplified to easily spot the issue -- i can make some bigger examples where it'd show up as suprising; e.g. using some `self.peers` in multiple steps of a quorum calculation, while in the middle of doing so someone might have reentrantly changed our set of peers etc.

Author: A more realistic example perhaps:

```
actor Acceptor { 
var peers: [Peer]

  func quorum(...) async -> Bool { 
    let oks: [Prepare] = []
    for peer in peers {
      await oks.append(peer.prepare(...))
    }
    log.info("Got OKs:\(oks) from \(peers.count)") // could be wrong "wrong" peers count

    guard oks count > (peers.count / 2 + 1) else { return false }

    for peer in peers {
      await peer.accept(...)
    }

    // implicit assumption that peers did not change between the runs?
  }
}
```

One would need to copy the peers to function scope to prevent them from accidentally be changed I guess...

It's a bit fuzzy to find a great example... should I try to find better ones?

With this example we have showcased that reentrant actors, by design, do not prevent "high-level" race conditions. They only prevent "low-level" data-races, as in: concurrent access to some variable they protect etc.

Please note that the example here a simple single variable state, yet already it can catch people off guard because how similar it looks to synchronous code. The problem is greatly exaggerated in real applications where actors are used to isolate and protect complex state transitions. Because actors effectively can only communicate through async calls to other parts of the system, a typical actor function will not have one or two suspension points but many of them - leading to exponential growth of the possible states an actor function may observe during it's execution.

> Existing actor implementations err on the side of non-reentrant actors by default, allowing for reentrancy to be optionally opted-into.
>
> - Erlang/Elixir ([gen_server](https://medium.com/@eduardbme/erlang-gen-server-never-call-your-public-interface-functions-internally-c17c8f28a1ee)) - showcases a simple "loop/deadlock" scenario and how to detect and fix it,

> > - Orleans ([Grains](https://dotnet.github.io/orleans/docs/grains/reentrancy.html)) offer extensive declarative configuration around reentrancy, it is also the closest to the Swift actors model,

Author: https://dotnet.github.io/orleans/docs/grains/reentrancy.htm - This is a very good read on the same issue and their solutions there; it matches up with how our actors look and behave like pretty well, so I'd recommend giving it a look

Reviewer: Orleans seems to allow recursive re-entrancy, at least.

Author: Yeah, I think it fits our model as well as basically by saying "within the same task" reentrance can happen which is like their "Reentrancy within a call chain".

Let me add this to our proposal as well

> Alternate spelling proposals follow below the example.

The rationale to make actors non-reentrant by default is clear: it is what feels natural and what actually enables "write as if synchronous code in an actor, and it just works" style of programming.

However, there are valid and important cases where we _do_ want to enable reentrant calls - e.g. a high priority message changing how we are processing a long running task inside an actor:

```swift
actor class ImageDownloader { 
  var bestImages: [Image]
  var currentBest: Image?
  
  func downloadAndPickBest(n: Int, urls: [URL]) async -> [Image] {
    for url in urls {
      let image = await download(url)
      bestImages.append(image)
      let ranking = await rank(image) 
      if ranking.isBest { 
        bestImage = image  
      }
      // ... more things to compute only the "best n" etc...
    }

    return images
  }
  
  @alwaysInterleave // ✅ may be invoked at any point in time 
  func currentBest() async -> Image { 
    currentBest

>   } 

Reviewer: I don't see how this is consistent with what you're saying.  If programs can just add an `@alwaysInterleave` method to an actor, then every other async function on that actor still has to worry about interleaving.  The idea that's consistent with your argument is that async functions have to consent to being interleaved with and otherwise run atomically.

Author: Hm yes... this allows for throwing in a wrench and "always interleave anyway" kind of intended for readonly queries or cancellation-like things mostly.

Your proposal makes sense, I guess it would be a `@reentrant func` then, which allows to be interleaved with.

Author: I rephrased the proposal to focus on specific functions opting into being reentrant, or the full actor opting in via `@reentrant` annotations.

I kept a mention of `@interleave(readOnly/unsafe)` but am not sure if we'd even want it (and express such doubts in the writeup)

}
```

> The above illustrates a popular use-case for reentrant calls: read only calls. They can be used to observe progress, request a "best effort" answer while better answers are being processed still, or one can invoke a cancel function to cancel some ongoing work inside an actor from another one.

Reviewer: An interleaved read can still read a value that the actor didn't intend for code to read.

Author: Right true, let me focus a bit more on @reentranc actor allowing to be interleaved with, rather this one. Thanks!

}
```

### Actor reentrancy (discussion)

One critical point that needs to be discussed and fleshed out is whether actors are [reentrant](https://en.wikipedia.org/wiki/Reentrancy_(computing)) by default or not.

The notion of reentrancy allows the actor runtime to claim the complete elimination of deadlocks, offers opportunity for scheduling optimization techniques where a "high priority task" _must_ be executed as soon as possible, however at the cost of _interleaving_ with any other actor-isolated function's execution any other asynchronous function declared on this actor. This imposes a large mental burden on developers, as every single actor function currently has to be implemented keeping reentrancy in mind, which effectively fails on delivering the promise of pain-free concurrency for the majority of use-cases.

> Currently, this proposal takes the aggressive approach of assuming _all_ actors are reentrant and not providing ways to opt out of this behavior. This section aims to highlight issues, benefits and tradeoffs with this approach as well as non-reentrancy in general. The goal of discussing these reentrancy issues, is to arrive at a design that is developer friendly, non-surprising, and delivers more completely on Swift's promise of _safe_ and _sane_ concurrency by default thanks to the use of actors.

Reviewer 2: ```suggestion
Currently, this proposal takes the aggressive approach of assuming _all_ actors are reentrant and not providing ways to opt out of this behavior. This section aims to highlight issues, benefits and tradeoffs with this approach as well as non-reentrancy in general. The goal of discussing these reentrancy issues, is to arrive at a design that is developer friendly, non-surprising, and delivers more completely on Swift's promise of _safe_ and _pain-free_ concurrency by default thanks to the use of actors.
```

Author: Ah, good one, thank you.

#### Reentrant actors: "Interleaving" execution

Reentrancy means that execution of asynchronous actor-isolated functions may "interleave" at suspension points, leading to increased complexity in programming with such actors, as every suspension point must be carefully inspected if the code _after_ it depends on some invariants that could have changed before it suspended.

> Interleaving executions still respect the actor's "single-threaded illusion"–i.e. no two functions will ever execute *concurrently* on any given actor–however they may _interleave_ at suspension points. In broad terms this means that reentrant actors are _thread-safe_ but are automatically protecting from the "high level" kinds of races that may still occur, potentially invalidating invariants upon which an executing asynchronous function may be relying on.

Reviewer 2: ```suggestion
Interleaving executions still respect the actor's "single-threaded illusion"–i.e. no two functions will ever execute *concurrently* on any given actor–however they may _interleave_ at suspension points. In broad terms this means that reentrant actors are _thread-safe_ but are not automatically protecting from the "high level" kinds of races that may still occur, potentially invalidating invariants upon which an executing asynchronous function may be relying on.
```

Some runtimes solve this by making *every single actor call have a timeout*. This would mean that each await could potentially throw, and that either timeouts or deadlock detection would have to always be enabled - which would be prohibitively expensive since we envision actors being used in the vast majority of concurrent Swift applications. It would also muddy the waters with respect to cancellation, which intentionally is designed to be explicit and cooperative, and as checking timeouts/deadlines is a form of cancellation, this is _not_ something we are going to support transparently, thus actor calls neither may assume this. 

It is easy to point out a small mistake in actors spanning a few lines of code, however programming complex actors with reentrancy can be quite a challenge. In this specific example, the solution–in hindsight–is simple, we should store the opinion in a function local variable, or in other words, any state the actor needs to to complete an execution "atomically" it must copy into local function scope. This can be hard to remember and manage consistently.

> > Depending on one's viewpoint, one could actually claim that deadlocks are better (!), than interleaving because they can be detected more easily and fixed more consistently.

Reviewer 2: ```suggestion
> Depending on one's viewpoint, one could actually claim that deadlocks are better (!), than interleaving because they can be reliably detected and explained by tools, can be detected in fuzz tests (no need to know what the correct result is for a random input), and can be fixed more consistently.
```

#### Reentrancy and async lets

Swift also offers the `async let` declaration style, allowing for expressing structured bounded number of asynchronous child tasks being performed concurrently.

> In order to sanity check our assumptions, let us also write some code using `async let` and see how reentrancy does or does not come into play here:

Reviewer 2: ```suggestion
In order to check our assumptions, let us also write some code using `async let` and see how reentrancy does or does not come into play here:
```

```swift
actor class Friend {
  func howMuchDoYouNeed() async -> Amount { ... }
  func send(cash: Amount) async { ... }
}

actor class Wallet {
  let friend: Friend = ... 
  
  var amount: Amount = ... 
  func checkCash() async -> Amount { ... }
  
  func lendFriendSomeCash() async  {
    async let requested = friend.howMuchDoYouNeed() 
    async let debt = checkDebt()
    async let cash = checkCash()
    let available = await cash - debt.amount
    
    if await requested <= available { 
      amount -= requested
      await friend.send(cash: requested) // ✅        
    }  
  }

  func looseWallet() async { 
    amount = 0
  }
}
```

This example composes well; Calls to an actor's `self` are allowed to pass through as usual, and _replies_ from other actors are also accepted. What can _not_ happen with non-reentrant actors is the `looseWallet()` function being randomly triggered while we are attempting to lend our friend some cash -- this would have been an external call into the actor, which our non-reentrancy rule would prevent.

So even such snippet (under non-reentrancy rules):

```swift
await wallet.lendFriendSomeCash()

> await wallet.looseWallet() 

Reviewer 2: ```suggestion
await wallet.loseWallet() 
```
(and everywhere else)

Author: Thanks, that word always gets me in english!

Assuming the non-reentrant actors as just discussed, it is still important to recognize that a frequent and usually quite understandable way of interacting between actors which are simply "conversations" between two or more actors in order fo fulfil some initial request.

Thanks to Swift embracing [Structured Concurrency](https://github.com/DougGregor/swift-evolution/blob/structured-concurrency/proposals/nnnn-structured-concurrency.md) as a core building block of it's concurrency story, we are in good position to do _better_ than just outright banning reentrancy. In Swift, every asynchronous operation is part of a `Task` which encapsulates the general computation taking place, and every asynchronous operation spawned from such task becomes a child Task of the current task. Synchronous calls do not change the current task (in a way, one can think of Tasks as similar to Threads, however they are not directly mapped to one another). Using this core capability in Swift's concurrency model, we are able to make actor calls *task-chain aware* and *allow* such calls to be reentrant.

> This resolves both the deadlock we encountered in the `Waiter` and `Kitchen` example in the previous section, and even enables implementing mutually recursive actors. In the following–world's silliest isEven implementation, we can see two actors performing either is even/odd checks, and mutually calling out to each other, because of the structured nature of tasks, and task awareness of the actor runtime, such calls would _not (!)_ deadlock under the task-aware runtime: 

Reviewer 3: Could you do the waiter kitchen example again and show the difference ? The odd even example doesn’t make it clear to me what is going on. Thanks!

Author: Thanks, will follow up on this

```swift
// WARNING: Don't actually implement an isOdd/isEven check like this, 
//          it involves multiple executor hops and is therefore very sub-optimal.
public actor class OddOddy { 
  let evan: EvenEvan!
  
  func isOdd(n: Int) async -> Bool {
    if n == 0 { return true }
    return await evan.isEven(num - 1)
  }
}

actor class EvenEvan {
  let oddy: OddOddy!

  func isEven(n: Int) async -> Bool {
    if n == 0 { return false }

>     return await oddy.isOdd(num - 1)

Reviewer 2: Do I understand correctly that an `async let` would start a new unrelated task that would not be allowed to interleave? If so, that's non-intuitive. I think it should be always correct to refactor an `await` expression into an `async let` that is immediately awaited:

```
    return await oddy.isOdd(num - 1)
```

===>

```
    async let result = oddy.isOdd(num - 1)
    return await result
```

Author: an `async let` does not introduce a suspension point, as such it does not offer a chance for interleaving (in any model); the interleaving could only happen at lines of code demarcated by `await` even if we had fully re-entrant actor only.

Reviewer 2: My point is that before the refactoring, we had only one parent task (that was executing both isEven and isOdd functions). After the refactoring, `async let` introduces a task for isOdd. Under the proposed task-aware runtime rules, it seems to me that these two pieces of code would be treated differently.

#### Reentrancy Summary

Preventing reentrancy complicates the model slightly, as there are cases where deadlocks can happen, however the gained benefit of an actor _truly_ being an a domain in which external calls are linearized and handled one after the other. Thanks to non-reentrant actors we can think of them as collections of small programs (their async functions triggered externally) which are triggered and run to completion, and then handle the next task, which greatly simplifies the mental model when working with those.

So, from a consistency point of view, one might want to prefer non-reentrant actors, but from an high-priority work scheduling in the style of "run this now, at any cost" reentrant actors offer an useful model, preventing data-races, while allowing this interleaved execution which–if one is *very careful*–can be utilized to some benefit.

> ---

Author: TODO: needs more cleanup from below here still, getting to it later tonight
